filament/backend/src/opengl/CocoaTouchExternalImage.mm

```diff
@@ -0,0 +1,311 @@
+/*
+ * Copyright (C) 2019 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "CocoaTouchExternalImage.h"
+
+#define GLES_SILENCE_DEPRECATION
+#include <OpenGLES/EAGL.h>
+#include <OpenGLES/ES3/gl.h>
+#include <OpenGLES/ES3/glext.h>
+
+#include "GLUtils.h"
+
+#include <math/vec2.h>
+
+#include <utils/Panic.h>
+
+namespace filament {
+
+static const char* s_vertexES = R"SHADER(#version 300 es
+in vec4 pos;
+void main() {
+    gl_Position = pos;
+}
+)SHADER";
+
+static const char* s_fragmentES = R"SHADER(#version 300 es
+precision mediump float;
+uniform sampler2D samplerLuminance;
+uniform sampler2D samplerColor;
+layout(location = 0) out vec4 fragColor;
+void main() {
+    float luminance = texelFetch(samplerLuminance, ivec2(gl_FragCoord.xy), 0).r;
+    // The color plane is half the size of the luminance plane.
+    vec2 color = texelFetch(samplerColor, ivec2(gl_FragCoord.xy) / 2, 0).ra;
+
+    vec4 ycbcr = vec4(luminance, color, 1.0);
+
+    mat4 ycbcrToRgbTransform = mat4(
+        vec4(+1.0000f, +1.0000f, +1.0000f, +0.0000f),
+        vec4(+0.0000f, -0.3441f, +1.7720f, +0.0000f),
+        vec4(+1.4020f, -0.7141f, +0.0000f, +0.0000f),
+        vec4(-0.7010f, +0.5291f, -0.8860f, +1.0000f)
+    );
+
+    fragColor = ycbcrToRgbTransform * ycbcr;
+}
+)SHADER";
+
+CocoaTouchExternalImage::CocoaTouchExternalImage(const CVOpenGLESTextureCacheRef textureCache)
+        noexcept : mTextureCache(textureCache) {
+
+    glGenSamplers(1, &mSampler);
+    glSamplerParameteri(mSampler, GL_TEXTURE_MIN_FILTER,   GL_NEAREST);
+    glSamplerParameteri(mSampler, GL_TEXTURE_MAG_FILTER,   GL_NEAREST);
+    glSamplerParameteri(mSampler, GL_TEXTURE_WRAP_S,       GL_CLAMP_TO_EDGE);
+    glSamplerParameteri(mSampler, GL_TEXTURE_WRAP_T,       GL_CLAMP_TO_EDGE);
+    glSamplerParameteri(mSampler, GL_TEXTURE_WRAP_R,       GL_CLAMP_TO_EDGE);
+
+    GLint status;
+
+    mVertexShader = glCreateShader(GL_VERTEX_SHADER);
+    glShaderSource(mVertexShader, 1, &s_vertexES, nullptr);
+    glCompileShader(mVertexShader);
+    glGetShaderiv(mVertexShader, GL_COMPILE_STATUS, &status);
+    assert(status == GL_TRUE);
+
+    mFragmentShader = glCreateShader(GL_FRAGMENT_SHADER);
+    glShaderSource(mFragmentShader, 1, &s_fragmentES, nullptr);
+    glCompileShader(mFragmentShader);
+    glGetShaderiv(mFragmentShader, GL_COMPILE_STATUS, &status);
+    assert(status == GL_TRUE);
+
+    mProgram = glCreateProgram();
```

Comment:
Yeah good point, I should make these static

Reply:
maybe not static because then they would be shared between different instances of filament -- this could work, but  makes things hard when shutting down.
maybe stash them in the main backend context/class.